src/main/java/org/broadinstitute/hellbender/dev/tools/walkers/bqsr/ApplyBQSRDataflow.java

```diff
@@ -0,0 +1,153 @@
+
+    @ArgumentCollection
+    public final RequiredReadInputArgumentCollection readArguments = new RequiredReadInputArgumentCollection();
+
+    @ArgumentCollection
+    protected IntervalArgumentCollection intervalArgumentCollection = new OptionalIntervalArgumentCollection();
+
+    /**
+     * Output path. Can be local or gs://
+     */
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc="Write output to this file")
+    public String OUTPUT;
+
+    /**
+     * Enables recalibration of base qualities.
+     * The covariates tables are produced by the BaseRecalibrator tool.
+     * Please be aware that you should only run recalibration with the covariates file created on the same input bam(s).
+     */
+    @Argument(fullName="bqsr_recal_file", shortName="bqsr", doc="Input covariates table file for base quality score recalibration")
+    public File BQSR_RECAL_FILE;
+
+    /**
+     * command-line options to fine tune the recalibration.
+     */
+    @ArgumentCollection
+    public ApplyBQSRArgumentCollection bqsrOpts = new ApplyBQSRArgumentCollection();
+
+    private SAMFileHeader header;
+
+
+    protected void setupPipeline(Pipeline pipeline) {
+        if (readArguments.getReadFilesNames().size()>1) {
+            throw new UserException("Sorry, we only support a single input file for now.");
+        }
+        String filename = readArguments.getReadFilesNames().get(0);
+        PCollection<Read> input;
+        try {
+            input = ingestReadsAndGrabHeader(pipeline, filename);
+        } catch (IOException x) {
+            throw new UserException.CouldNotReadInputFile(filename, x);
+        }
+        // Surely we'll be using this commented-out code once ReadsSource takes stringency into account.
+        //ReadsSource readsSource = new ReadsSource(filename, pipeline);
+        //SAMFileHeader header = readsSource.getHeader();
+        //final SAMSequenceDictionary sequenceDictionary = header.getSequenceDictionary();
+        //final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(sequenceDictionary) :
+        //        IntervalUtils.getAllIntervalsForReference(sequenceDictionary);
+        final BaseRecalOutput recalInfo = new BaseRecalOutput(BQSR_RECAL_FILE);
+        PCollection<BaseRecalOutput> recalInfoSingletonCollection = pipeline.apply(Create.of(recalInfo).withName("recal_file ingest"));
+        PCollection<Read> output = input // readsSource.getReadPCollection(intervals, ValidationStringency.SILENT)
+           .apply(new ApplyBQSRTransform(header, recalInfoSingletonCollection, bqsrOpts));
+        SmallBamWriter.writeToFile(pipeline, output, header, OUTPUT);
+    }
+
+    /** reads local disks or GCS -> header, and PCollection */
+    private PCollection<Read> ingestReadsAndGrabHeader(final Pipeline pipeline, String filename) throws IOException {
```

Comment:
yup. This isn't an issue anymore now that I'm using a "clean" input, though, so I was able to go back to ReadSource.